### ISSUE TITLE
E-invoice for Europe to comply with new EU directive

### ISSUE THREAD

**Issue by approde** [InvoiceShelf/InvoiceShelf#163]

**Comment by rihards-simanovics** [comment]
@approde please provide more info.

**Comment by approde** [comment]
https://en.e-rechnung-bund.de/ 

You can find more information here. I am also willing to donate for this function.

**Comment by approde** [comment]
https://en.e-rechnung-bund.de/ 

You can find more information here. I am also willing to donate for this function.

Its a future request

**Comment by rihards-simanovics** [comment]
I think this is something we already discussed in the discord it's on the radar of the maintainers, thanks for the input 😊!

**Comment by approde** [comment]
@rihards-simanovics  Oh sorry, I didn't know that. Thank you very much for your feedback!

**Comment by approde** [comment]
@rihards-simanovics  Oh sorry, I didn't know that. Thank you very much for your feedback

**Comment by rihards-simanovics** [comment]
For clarity here is the link to the discussion, if you are on our discord server you can see it there: [🆕feature-suggestions: European union directive for E-Invoices](https://discord.com/channels/1200813332551192606/1284798704162836541)

**Comment by rihards-simanovics** [comment]
Re-opening as a feature request

**Comment by Brovning** [comment]
Hello,

sorry, I haven't seen, that there is already a similar ticket, because I have searched for "ZUGFeRD"...

Let me add my information as well here:

> What is ZUGFeRD?
ZUGFeRD is a cross-industry data format for electronic invoice data exchange that was developed by the Forum for Electronic Invoicing Germany (FeRD) - with support from the Federal Ministry of Economics and Energy. The ZUGFeRD data format is based on the Directive 2014/55/EU of 16 April 2014 on electronic invoicing for public contracts and on the EN16931 standard published on 28 June 2017. Also, the Cross-Industry Invoice (CII) of UN/CEFACT and the ISO standard 19005-3:2012 (PDF/A-3) are also taken into account as of ZUGFeRD 2.0.
The areas of application for this hybrid invoice format are diverse. ZUGFeRD can be used in B2B, B2G, and B2C business transactions.

Description of the Standard:
https://www.ferd-net.de/standards/zugferd-2.3/zugferd-2.3.html

**Comment by rihards-simanovics** [comment]
don't worry @Brovning I closed it as the initial message was a question, but I converted it to a Feature Request.

**Comment by McNetic** [comment]
The EU directive requires compliance with regards to eletronic invoices only for B2B transactions; for these XRechung (https://xeinkauf.de/xrechnung/) is a much simpler and better suited format that is also standards compliant.

**Comment by Brovning** [comment]
Yes, that's right.
There are two standards which can be used, BUT:
ZUGFeRD is having the advantage that all information is within one PDF file.
With the XRechnung format you are having a separated XML, which is from my point of view a drawback.

Having ZUGFeRD implemented, you can send the PDF (containing the XML information) to everybody, independent if he likes to use the electronic invoice capabilities or not. For everybody else it is human readable, as it is a PDF file.

**Comment by rihards-simanovics** [comment]
Since this is Country/Region specific, a Module would be the best way to move forward with it.

**Comment by McNetic** [comment]
@Brovning I agree that ZUGFeRD seems to be a better choice for B2C invoices on the first glance.

It has, however, one fatal and incorrectible flaw: there is and will be no way to ensure the invoice data rendered in the PDF part of the document is identical to the embedded XML part. This is no problem as long as there are no malicious parties involved. There is also, and while not having electronic invoice capabilities, will never be, a way for the recipient of the ZUGFeRD document to check if the two parts are identical. Moreover, it will be technically challenging, if not impossible, to verify the two parts are identical by technical means.

On the other hand, automatic processing of the invoice documents will primarily rely on the XML part, so a big source of errors arises, especially for the non tech-save clients without special software. As soon as the electronic invoices gain traction in the B2C environment, malicious parties will pop up from everywhere, trying to trick users to accept invoices with malicious data.

But back to topic: I mainly promote the XRechnung for B2B, because here, the technical ability to accept ANY standards compliant format is required for every european business by law beginning next year. So if you have to use a software that can read any electronic invoice anyways, the PDF part becomes usesless and you will never look at anything besides the XML part, anyways.

**Comment by Brovning** [comment]
There is a PHP SDK for ZUGFeRD/XRechnung:
https://github.com/easybill/zugferd-php

> ZUGFeRD PHP SDK (Factur-X, XRechnung) - Convert PHP Objects to XML and back.

Tests are available:
https://github.com/easybill/zugferd-php/tree/master/tests/zugferd10/Tests

@rihards-simanovics
Is this something we can start with?

**Comment by McNetic** [comment]
> With the XRechnung format you are having a separated XML, which is from my point of view a drawback.
With XRechnung, you only have an XML file. There is nothing else.

**Comment by Brovning** [comment]
> > With the XRechnung format you are having a separated XML, which is from my point of view a drawback.
> 
> With XRechnung, you only have an XML file. There is nothing else.

Which is a major disadvantage for me and most of my customers...

**Comment by rihards-simanovics** [comment]
Looks to me as though it's not very simple, I'm going to convert this into a discussion so we can probe all the sides of this before going for implemintation.
